core/src/main/java/software/amazon/awssdk/auth/profile/internal/ProfileFileLocations.java

```diff
@@ -0,0 +1,93 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.auth.profile.internal;
+
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+import java.util.regex.Pattern;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.core.AwsSystemSetting;
+import software.amazon.awssdk.utils.JavaSystemSetting;
+import software.amazon.awssdk.utils.StringUtils;
+
+/**
+ * A collection of static methods for loading the location for configuration and credentials files.
+ */
+@SdkInternalApi
+public final class ProfileFileLocations {
+    private static final Pattern HOME_DIRECTORY_PATTERN =
+            Pattern.compile("^~(/|" + Pattern.quote(FileSystems.getDefault().getSeparator()) + ").*$");
+
+    private ProfileFileLocations() {}
+
+    /**
+     * Load the location for the configuration file, usually ~/.aws/config unless it's overridden using an environment variable
+     * or system property.
+     */
+    public static Optional<Path> configurationFileLocation() {
+        return AwsSystemSetting.AWS_CONFIG_FILE.getStringValue()
+                                               .flatMap(ProfileFileLocations::resolveProfileFilePath);
+    }
+
+    /**
+     * Load the location for the credentials file, usually ~/.aws/credentials unless it's overridden using an environment variable
+     * or system property.
+     */
+    public static Optional<Path> credentialsFileLocation() {
+        return AwsSystemSetting.AWS_SHARED_CREDENTIALS_FILE.getStringValue()
+                                                           .flatMap(ProfileFileLocations::resolveProfileFilePath);
+    }
+
+    /**
+     * Load the home directory that should be used for the profile file. This will check the same environment variables as the CLI
+     * to identify the location of home, before falling back to java-specific resolution.
+     */
+    public static String userHomeDirectory() {
+        boolean isWindows = JavaSystemSetting.OS_NAME.getStringValue()
+                                                     .map(s -> StringUtils.lowerCase(s).startsWith("windows"))
+                                                     .orElse(false);
+
+        // To match the logic of the CLI we have to consult environment variables directly.
+        // CHECKSTYLE:OFF
+        if (System.getenv("HOME") != null) {
+            return System.getenv("HOME");
+
+        } else if (isWindows && System.getenv("USERPROFILE") != null) {
```

Comment:
minor, but we can probably remove all the `else`s, it might be cleaner that way and I would make those env strings constants since all of them are used twice here and it's hard to tell if there's a mismatch or a typo.

Reply:
+1

Reply:
I like separating the ifs, but the constants made the code harder to read at a glance. A reader has to jump up to the top of the file to see what the actual environment variable is and then jump back down. The unit tests make sure we don't have typos. I've extracted out the repeated `getenv` calls as a potential compromise?

Reply:
👍